pkg/synchronization/core/scan.go

```diff
@@ -326,6 +333,11 @@ func (s *scanner) directory(
 			directory = d
 			defer directory.Close()
 		}
+	} else if directory != nil && s.ignorerMode == "dockerignore" {
+		// If directory is not nil, then then is the start of a scan
+		// so we will use the directory walk logic that is compatible
+		// with dockerignore syntax.
+		return s.directoryWithDockerIgnoreCompatibility(filepath.Join(s.root, path), metadata, directory, baseline)
```

Comment:
I'd prefer to avoid a completely separate directory scanning implementation (though I understand why that's done currently). It really would be better if this could just be handled in the existing directory function with only different ignore behavior (i.e. lines [394-406](https://github.com/mutagen-io/mutagen/blob/af15c085b40d410426c3e80d7d978cdef35f2e50/pkg/synchronization/core/scan.go#L394-L406)). That would avoid the need for extracting the content handling logic into a separate function.

Reply:
Yeah, I think it is possible to do as you suggest.  One thing we will need to sort out is the "partial match" concept, maybe this is as simple as adding a new bool value to the response of `directory`.  Since we will need to recursively scan on patterns like `**/*.go` we will need to prune out the directory scan results if there are ultimately no matches found (otherwise we end up with a directory tree with all empty directories).  This is what `fsutil` does for us, but I think won't be to hard to reproduce that logic with docker `fileutils` as you mention.